cEP-0028.md

```diff
@@ -0,0 +1,445 @@
+     provides a mechanism to handles duplicates
+     and resolving them.
+     """
+
+     def __init__(self, file_proxies=[]):
+       """
+       :param file_proxies:
+           A list of FileProxy instances to initialize
+           the ProxyMap with.
+       """
+       self._map = {proxy.filename: proxy for proxy in file_proxies}
+
+     def add(self, proxy: FileProxy, replace=False):
+       """
+       Add a proxy instance to the proxy map.
+
+       :param proxy:
+           The proxy instance to register in the map.
+       :param replace:
+           A boolean flag indicating if the proxy should
+           replace an existing proxy of the same file.
+       :return:
+           Boolean true if registering of the proxy was
+           successful else false.
+       """
+       if not isinstance(proxy, FileProxy):
+         return False
+
+       if self._map.get(proxy.filename) is not None:
+         if replace:
+           self._map[proxy.filename] = proxy
+           return True
+         return False
+
+       self._map[proxy.filename] = proxy
+       return True
+
+     def resolve(self, filename: str, workspace=None, hard_sync=True):
+       """
+       Resolve tries to find an available proxy or creates one
+       if there is no available proxy for the said file.
+
+       :param filename:
+           The filename to search for in the map or to create
+           a proxy instance using.
+       :param workspace:
+           Used in case the lookup fails and a new instance is
+           being initialized.
+       :hard_sync:
+           Boolean flag indicating if the file should be initialized
+           from the file on disk or fail otherwise.
+       :return:
+           Returns a proxy instance or a boolean indicating the
+           failure of the resolution.
+       """
+       proxy = self.get(filename)
+       if proxy is not None:
+         return proxy
+
+       try:
+         proxy = FileProxy.from_name(filename, workspace)
+       except FileNotFoundError:
+         if hard_sync:
+           return False
+
+       try:
+         proxy = FileProxy(filename, workspace)
+       except Exception:
+         return False
+
+       self.add(proxy)
+       return proxy
+   ```
+
+   - `coalib.processes.Processing.get_file_dict_from_fileproxy_map()`
+
+     Requires addition of the method to replace `get_file_dict()` when
+     coala API is called with a FileProxyMap instance.
+
+   ```python
+   def get_file_dict(filename_list,
+                     fileproxy_map,
+                     log_printer=None):
+     """
+     This is the function that is intended to patch the existing
+     implementation of coalib.processes.Processing.get_file_dict().
+     Reads all files into a dictionary.
+
+     :param filename_list:   List of names of paths to files to get contents of.
+     :param fileproxy_map:   FileProxyMap containing the files available.
+     :param log_printer:     The logger which logs errors.
```

Comment:
we dont use `log_printer` any more

Reply:
Remove it?

Reply:
hmmm... how are you 'patching' this?

If it is going to be an additional function, then it shouldnt have the same name as the main `get_file_dict` ?

Reply:
The method name has to be different, I will fix it. The current idea is to conditionally use `get_file_dict_from_fileproxy_map()` instead of `get_file_dict()` depending upon the presence of a fileproxy map in `instantiate_processes()`.